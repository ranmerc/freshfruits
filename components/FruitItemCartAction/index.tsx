import { useCartContext } from "@/context/CartContext";
import PriceData from "@/types/PriceData";
import { AddCircle, RemoveCircle } from "@mui/icons-material";
import {
  Button,
  FormControl,
  IconButton,
  InputLabel,
  MenuItem,
  Select,
  Stack,
  Typography,
} from "@mui/material";
import React, { useMemo } from "react";
import calculateDiscount from "@/utils/calculateDiscount";

export default function FruitItemCartAction({
  prices,
  id,
}: {
  prices: PriceData;
  id: number;
}) {
  const [selectedPack, setSelectedPack] = React.useState(0);
  const { cartItems, addItemToCart, removeItemFromCart } = useCartContext();

  const quantityInCart = useMemo(() => {
    const item = cartItems.find(
      (item) => item.fruitId === id && item.selectedPackId === selectedPack
    );
    return item?.quantity || 0;
  }, [cartItems, id, selectedPack]);

  const itemToBeAdded = useMemo(() => {
    return {
      fruitId: id,
      pack: prices.packs[selectedPack],
      quantity: 1,
      selectedPackId: selectedPack,
    };
  }, [id, selectedPack, prices]);

  return (
    <>
      <Stack rowGap={2}>
        {prices.packs[selectedPack].price !==
          prices.packs[selectedPack].discountPrice && (
          <Typography
            variant="caption"
            sx={{
              bgcolor: "secondary.main",
              padding: 0.5,
              fontSize: "0.8rem",
              textAlign: "center",
            }}
          >
            {calculateDiscount(
              prices.packs[selectedPack].discountPrice,
              prices.packs[selectedPack].price
            )}{" "}
            % off
          </Typography>
        )}
        <FormControl fullWidth>
          <InputLabel id={`packs-for-${id}`}>Pack</InputLabel>
          <Select
            labelId={`packs-for-${id}`}
            id={`select-packs-for-${id}`}
            value={selectedPack}
            label="Pack"
            onChange={(e) => {
              setSelectedPack(e.target.value as number);
            }}
          >
            {prices.packs.map((pack, i) => {
              if (pack.inStock) {
                return (
                  <MenuItem value={i} key={i}>
                    {pack.quantity} {pack.type === "count" ? "Nos" : "gm"} for â‚¹
                    {pack.discountPrice}
                  </MenuItem>
                );
              }
            })}
          </Select>
        </FormControl>

        {quantityInCart !== 0 ? (
          <Stack
            direction={"row"}
            alignItems={"center"}
            sx={{
              bgcolor: "secondary.main",
              borderRadius: 2,
              justifyContent: "space-around",
            }}
          >
            <IconButton
              onClick={() =>
                removeItemFromCart({
                  fruitId: id,
                  selectedPackId: selectedPack,
                })
              }
              aria-label="Remove from Cart"
            >
              <RemoveCircle />
            </IconButton>
            <Typography fontSize={"1.2rem"}>{quantityInCart}</Typography>
            <IconButton
              onClick={() => addItemToCart(itemToBeAdded)}
              aria-label="Add to Cart"
            >
              <AddCircle />
            </IconButton>
          </Stack>
        ) : (
          <Button
            size="large"
            variant="outlined"
            onClick={() => addItemToCart(itemToBeAdded)}
          >
            Add
          </Button>
        )}
      </Stack>
    </>
  );
}
