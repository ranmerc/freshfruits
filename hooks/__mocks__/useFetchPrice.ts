const useFetchPrice = jest.fn(() => {});

export default useFetchPrice;
