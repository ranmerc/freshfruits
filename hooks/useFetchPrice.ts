import PriceData from "@/types/PriceData";
import axios from "axios";
import { useQuery } from "@tanstack/react-query";

const fetcher = async (id: number) => {
  const response = await axios.get<PriceData>(`/api/pricing?id=${id}`);
  return response.data;
};

export default function useFetchPrice(id: number) {
  return useQuery<PriceData>({
    queryKey: ["fruit", id],
    queryFn: () => fetcher(id),
  });
}
