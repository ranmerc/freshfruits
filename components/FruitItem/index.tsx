import useFetchPrice from "@/hooks/useFetchPrice";
import FruitData from "@/types/FruitData";
import {
  Alert,
  Card,
  CardActions,
  CardContent,
  CardMedia,
  Typography,
} from "@mui/material";
import Link from "next/link";
import FruitItemCartAction from "../FruitItemCartAction";

export default function FruitItem({ item }: { item: FruitData }) {
  const {
    data: priceData,
    isLoading,
    isError,
    isSuccess,
  } = useFetchPrice(item.id);

  if (isLoading) {
    return <Alert severity="info">Loading prices...</Alert>;
  }

  if (isError) {
    return (
      <Alert severity="error">Error fetching price data for {item.name}</Alert>
    );
  }

  return (
    <>
      <Card sx={{ p: 2, height: "100%" }}>
        <Link
          href={`/fruit/${item.id}`}
          style={{
            textDecoration: "none",
            color: "inherit",
          }}
        >
          <CardMedia
            image={item.images[0]}
            title={`${item.name} image`}
            sx={{ height: 150, width: "100%" }}
          />
          <CardContent sx={{ px: 0, py: 2 }}>
            <Typography
              variant="h4"
              fontWeight={"bold"}
              textTransform={"capitalize"}
              sx={{ textAlign: "center", overflowX: "scroll" }}
            >
              {item.name}
            </Typography>
          </CardContent>
        </Link>
        <CardActions>
          {isSuccess && priceData && priceData.packs.length !== 0 && (
            <FruitItemCartAction prices={priceData} id={item.id} />
          )}
        </CardActions>
      </Card>
    </>
  );
}
