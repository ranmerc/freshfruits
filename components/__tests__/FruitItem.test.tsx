import { screen, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import FruitItem from "../FruitItem";
import useFetchPrice from "@/hooks/useFetchPrice";
import FruitItemCartAction from "../FruitItemCartAction";
import renderer from "react-test-renderer";

jest.mock("@/hooks/useFetchPrice");
jest.mock("@/components/FruitItemCartAction", () => {
  return function FruitItemCartAction() {
    <div></div>;
  };
});

const mockedUseFetchPrice = useFetchPrice as jest.Mock;

describe("FruitItem component", () => {
  it("Renders loading state", () => {
    mockedUseFetchPrice.mockReturnValue({
      isLoading: true,
      isError: false,
      isSuccess: false,
      data: null,
    });

    const component = (
      <FruitItem
        item={{
          countryOfOrigin: "India",
          description: "Fruit Description",
          id: 1,
          images: ["images.jpg"],
          name: "Fruit Name",
          shelfLife: 7,
          storageInstructions: "Storage instructions",
        }}
      />
    );

    render(component);

    expect(screen.getByRole("alert")).toHaveTextContent("Loading prices...");

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Renders error state", () => {
    mockedUseFetchPrice.mockReturnValue({
      isLoading: false,
      isError: true,
      isSuccess: false,
      data: null,
    });

    const component = (
      <FruitItem
        item={{
          countryOfOrigin: "India",
          description: "Fruit Description",
          id: 1,
          images: ["images.jpg"],
          name: "Fruit Name",
          shelfLife: 7,
          storageInstructions: "Storage instructions",
        }}
      />
    );

    render(component);

    expect(screen.getByRole("alert")).toHaveTextContent(
      "Error fetching price data for Fruit Name"
    );

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Renders correct fruit details", () => {
    mockedUseFetchPrice.mockReturnValue({
      isLoading: false,
      isError: false,
      isSuccess: true,
      data: {
        id: 1,
        packs: Array(2),
      },
    });

    const component = (
      <FruitItem
        item={{
          countryOfOrigin: "India",
          description: "Fruit Description",
          id: 1,
          images: ["images.jpg"],
          name: "Fruit Name",
          shelfLife: 7,
          storageInstructions: "Storage instructions",
        }}
      />
    );

    render(component);

    expect(
      screen.getByRole("link", {
        name: "Fruit Name",
      })
    ).toHaveAttribute("href", "/fruit/1");

    expect(
      screen.getByRole("img", { name: "Fruit Name image" })
    ).toBeInTheDocument();

    expect(screen.getByRole("heading", { name: "Fruit Name" }));

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
