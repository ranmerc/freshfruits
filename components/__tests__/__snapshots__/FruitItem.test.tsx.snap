// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FruitItem component Renders correct fruit details 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-t4nkj9-MuiPaper-root-MuiCard-root"
>
  <a
    href="/fruit/1"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
    style={
      {
        "color": "inherit",
        "textDecoration": "none",
      }
    }
  >
    <div
      className="MuiCardMedia-root css-1kils9u-MuiCardMedia-root"
      role="img"
      style={
        {
          "backgroundImage": "url("images.jpg")",
        }
      }
      title="Fruit Name image"
    />
    <div
      className="MuiCardContent-root css-1fe2h86-MuiCardContent-root"
    >
      <h4
        className="MuiTypography-root MuiTypography-h4 css-1kxrs9j-MuiTypography-root"
      >
        Fruit Name
      </h4>
    </div>
  </a>
  <div
    className="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
  />
</div>
`;

exports[`FruitItem component Renders error state 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-1hdyuqk-MuiPaper-root-MuiAlert-root"
  role="alert"
>
  <div
    className="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
      data-testid="ErrorOutlineIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
      />
    </svg>
  </div>
  <div
    className="MuiAlert-message css-1pxa9xg-MuiAlert-message"
  >
    Error fetching price data for 
    Fruit Name
  </div>
</div>
`;

exports[`FruitItem component Renders loading state 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard css-1fuz9qr-MuiPaper-root-MuiAlert-root"
  role="alert"
>
  <div
    className="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
      data-testid="InfoOutlinedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
      />
    </svg>
  </div>
  <div
    className="MuiAlert-message css-1pxa9xg-MuiAlert-message"
  >
    Loading prices...
  </div>
</div>
`;
